[Refactoring] Use a generic error type instead of the macro `error_impls`

nepalez opened the pull request:
While trying to recognize how the gem works I found this [TODO](https://github.com/nats-io/nats.rs/blob/main/async-nats/src/lib.rs#L1346) and decided to implement it (well, this is how I learn things).

But at the moment the implementation got ready, I asked myself, why the same functionality cannot be expressed better with generic structs, and found no answer. That's why instead of implementing another macro (derive or attribute), I restarted my refactoring from scratch and removed the very necessity of the macro in favor of `NatsError<T>`.

This type represents any error to be raised by the crate. It is generic over the enum representing the kind of the error.

Unlike the previous code where binding between the kind and the error used to be provided by the macro `error_impls`, this definition makes the corresponding error bound to the kind more idiomatically.

The previous syntax:

```rust
#[derive(Clone, Debug, PartialEq)]
pub enum FooErrorKind {
  Bar,
  Baz,
}

#[derive(Debug)]
pub struct FooError {
  kind: FooErrorKind,
  source: Box<dyn Error + Send + Sync + 'static>,
}

error_impls!("FooError", "FooErrorKind");
```

The new definition:

```rust
#[derive(Clone, Debug, PartialEq)]
pub enum FooErrorKind {
  Bar,
  Baz,
}

pub type FooError = NatsError<FooErrorKind>;
```

In addition to being more expressive (which is, I agree, too opinionated), the new syntax lets IDE-s to understand the code better (most of them cannot analyze the code inside the macro).

Jarema commented:
Hey!

That's a very nice approach and one of the options we were debating as an internals refactor.

Will review soon.
Thank you!

caspervonb commented:
I'm for it (as previously discussed when you were working on the errors @Jarema 😁 )

nepalez commented:
I've rebased the branch over the fresh main branch and also fixed formatting along with a typo in the error description.

Jarema commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+        assert_eq!(error.format_source(), "unknown");
+    }
+
+    #[test]
```
I would remove this test.

nepalez replied:
Fixed

Jarema commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+use std::error::Error;
+use std::fmt::{Debug, Display, Formatter};
+
+// Notice: see unit tests to see how to use this generic error class.
```
Remove the notice.

nepalez replied:
Fixed

Jarema reviewed (changes requested):
In general, this is a really nice change, but there are a few nits to be more Rust-idiomatic and consistent with the rest of library.

Thanks a lot for the work you've done here!

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+// Notice: see unit tests to see how to use this generic error class.
+
+/// The error type for the NATS client, generic by the kind of error.
+pub struct NatsError<Kind>
```
I'm not sure about this. The reason is nats already have [the Error type](https://github.com/nats-io/nats.rs/blob/9677baf6dcbf018af16e7f1e295bdddf9d108666/async-nats/src/lib.rs#L150) which is a part of the public interface.

Maybe it is better to use something like `CustomError<T>` here to prevent echoing?

nepalez replied:
Alternatively, I could customize all errors [like this]() as a types of `NatsError`, so that the crate would only provide the `NatsError<T>`, and then make 2 renamings:
- `crate::Error` -> `crate::SourceError` (dynamic)
- `crate::NatsError` -> `crate::Error` (static)

Jarema replied:
The `crate::Error` was added just for having simple alias for typical Rust Boxed error with intention to be used like in all doc examples. It became irrelevant after introducing concrete errors for whole API.

Then it was used for `source`, to have less verbose and consistent syntax at each `source` field. This became irrelevant too, as your improvement gets rid of repeated structs with `source` field.

Changing `crate::Error` name would be a breaking change without good reason. I would prefer to deprecate it.

**Also, this new `NatsError` does not have to be public, as we make all aliases (`type Foo<FooKind>` public).
I was considering if it would "voldemort" the type, but it should be fine.**

Jarema replied:
Oh, and yes, we need to add concrete errors to `service` API. We want the whole API of this crate to be based of concrete errors.

caspervonb replied:
Tangentially, crate::Error should in the long term, be removed imo as explicitly saying what it is in say `main` makes for better documentation examples.

This type error type introduced here being public depends on how rustdoc renders it if it was voldemorted.

Jarema replied:
actually, it has to be public, otherwise we would get:

```
error[E0446]: crate-private type `nats_error::NatsError<jetstream::stream::ConsumerErrorKind>` in public interface
    --> async-nats/src/jetstream/stream.rs:1782:1
     |
1782 | pub type ConsumerError = NatsError<ConsumerErrorKind>;
     | ^^^^^^^^^^^^^^^^^^^^^^ can't leak crate-private type
     |
    ::: async-nats/src/nats_error.rs:6:1
     |
6    | pub(crate) struct NatsError<Kind>
     | --------------------------------- `nats_error::NatsError<jetstream::stream::ConsumerErrorKind>` declared as crate-private
```
     
But it does not have to be re-exported as `crate::Error`, but can remain `crate::error::Error`, while deprecating `crate::Error`.

  WDYT @caspervonb ?

Jarema replied:
To be fair, I would not overthink it.
Keep in mind it's an internal API, and often simple and explicit one is better than a smart one. Especially if you want to encourage contributions. The lower the library learning curve, the better :). 

Regarding kinds - this was deliberate decision:
We expose error enum variants only if it makes sense for users to match against them, which is when user can take some action because of error kind returned.

In case of those errors you linked as example it does not make sense for the source kind to exist as separate kind, nor as a kind on returning error level. It's just giving contextual information when logging, which is really easy with those literal &str's.

nepalez replied:
Ahh, I missed your comment yesterday. Anyway, I don't use those new methods `into()` and `with_kind()` so they can be dropped without much sadness ))

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+}
+
+impl<Kind> Debug for NatsError<Kind>
```
Fixed

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    #[test]
+    fn new_method_works() {
```
Fixed

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    #[test]
+    fn with_source_method_works() {
```
Fixed

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    #[test]
+    fn kind_method_works() {
```
Fixed

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    #[test]
+    fn format_source_method_works_with_source() {
```
Fixed

nepalez commented on async-nats/src/nats_error.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    #[test]
+    fn format_source_method_works_without_source() {
```
Fixed

nepalez commented:
I have added 3 big changes into the PR

- "normalized" errors so that the `source` is separated from the `kind` (previously `JetStream(err)` version of some kinds were keeping the source inside)
- added the default implementation of `Display` right to the error, and in the code moved `impl Display` to the kinds (this is becasue after the previous step this became a common pattern)
- implemented 2 methods out of the box to reflect the fact an error is defined by its kind:

```rust
FooErrorKind::SomeError.into() // instead of FooError::new(FooErrorKind::SomeError)
source.with_kind(FooErrorKind::SomeError) // instead of FooError::with_source(FooErrorKind::SomeError, source)
```

Jarema commented:
> I have added 3 big changes into the PR
> 
> * ["normalized" errors](https://github.com/nats-io/nats.rs/pull/1047/commits/9c33e2d2b8094e64c5b8a87badb151f62f2c8ec6) so that the `source` is separated from the `kind` (previously `JetStream(err)` version of some kinds were keeping the source inside)

That was done by purpose.

Users (and library itself) need a clean way to get to the underlying JetStream Error easily.
It does not look nice if user matches against kind, sees that it is `JetStream` Kind, yet they need to unwrap if `source` is `Some` and `downcast` error to get the JetStream Error. And the JetStream error is what in many cases user or we need to match against. It also adds possibility to introduce bugs by forgetting to add that source.

I know it's not perfect, but to be fair, JetStream errors on server side are not very clean, and new ones are added over time, so adding them directly to `Kind` would require all kinds to be `non_exhaustive` plus it's really hard to find all possible JS errors that can happen for given API call (to not call it impossible).

nepalez commented:
@Jarema thank you for the explanation. Well, I need to think a bit more about it. Maybe it would block the generalization of the `Display`.

Another question is about your previous comment. As far, as I understood, you're not a big fan of making all errors wrapped and standardized in this way? To be honest, I'm not experienced enough for rust codebases, so I have no any opinion about what approach is preferable in Rust.

In Ruby world where I'm from, I always prefer the library to provide its own errors, even though it is just a thin wrapper around some third-party ones. This simplifies the debugging by pointing out to the specific package. Following this approach, I could potentially wrap all `std::io::Error`-s into the corresponding `crate::Error`, but **should I**?

nepalez commented:
I've dropped the commit and returned to the `JetStream(err)`. The implementation of the `Display` trait survived :)

Jarema commented:
> In Ruby world where I'm from, I always prefer the library to provide its own errors, even though it is just a thin wrapper around some third-party ones. This simplifies the debugging by pointing out to the specific package. Following this approach, I could potentially wrap all `std::io::Error`-s into the corresponding `crate::Error`, but **should I**?

It's the same in Rust - you usually do not want to expose the types of dependencies of your library.
However, `JetStreamError` IS a type from this crate :).

For learning context: There are exceptions though.  One of few non-library types you can see exposed in the wild a lot is `std::io::Error`.

Jarema commented:
@nepalez Looks good now! Rename the file and error and it's LGTM :).

nepalez commented:
@Jarema done ))

caspervonb commented on async-nats/src/error/mod.rs:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    // In some cases the kind doesn't implement `Copy` trait
+    pub fn kind(&self) -> Kind {
```
This can just return ref? no clone needed.

```suggestion
    pub fn kind(&self) -> &Kind {
```

nepalez replied:
I asked myself the same question, but left this method for the backward compatibitlity only. Without it we could just make the `kind` attribute public ))

Jarema replied:
Its Copy most of the time. Clone only in case or JetStream error. And if you look at std::io::Error its also returning copy.

nepalez replied:
Not only the `JetStream`, but also a Request keeping a string inside, which is not copy.

caspervonb reviewed (commented):
Do we want this to be a ref instead of cloning? cc @Jarema

Jarema reviewed (approved):
LGTM!

Jarema reviewed (changes requested):
Oh, that LGTM was a little premature.

Not sure for reason on implementing Display on Kinds, instead of Error?
That way, we loose quite a lot of useful context,

like here:
https://github.com/nats-io/nats.rs/pull/1047/files#diff-09d2b3ba6ced37b50fa193faa24082492564d93fffcf3068af4ee69a4be7e54dL884-R862

or here:
https://github.com/nats-io/nats.rs/pull/1047/files#diff-014b24d4d8bf5b62ffeed0269754925ee12069d078d371058be69986cb926ea8L963-L972

Plus, the default display rather hurts than helps - without it, if you add new error compiler will remind you that you need a nice display impl for it. Without it - it's very easy to skip it and have the very basic one, which is too basic in my opinion.

Sorry for all those change requests.
That's why we prefer iterative work with smaller PRs. Much easier to merge in changes that way :).

nepalez commented:
> we loose quite a lot of useful context,

I believe we don't because this context [is still available](https://github.com/nats-io/nats.rs/pull/1047/files#diff-dd98d7aacc9e5caae85892fe7caba5ce8e3d9e5b520d4a79acd40e278a02e4b8R37-R48) in the error's `Display`.

As long as the error has a source, this context is added to the error.

The only thing I am not sure about is this. For now all kinds [require](https://github.com/nats-io/nats.rs/pull/1047/files#diff-dd98d7aacc9e5caae85892fe7caba5ce8e3d9e5b520d4a79acd40e278a02e4b8R7) `Display` to be implemented. I could loose this requirement here and let the developer to decide if he is going to imlement `Display` for the kind or for the whole error. Only if he implements the `Display` for the kind, the error would be displayed as well out of the box.

Jarema commented:
> > we loose quite a lot of useful context,
> 
> I believe we don't because this context [is still available](https://github.com/nats-io/nats.rs/pull/1047/files#diff-dd98d7aacc9e5caae85892fe7caba5ce8e3d9e5b520d4a79acd40e278a02e4b8R37-R48) in the error's `Display`. As long as the error has a source, its description is added to the error.
> 
> The only thing I am not sure about is this. For now all kinds [require](https://github.com/nats-io/nats.rs/pull/1047/files#diff-dd98d7aacc9e5caae85892fe7caba5ce8e3d9e5b520d4a79acd40e278a02e4b8R7) the `Display` to be implemented for them.
> 
> I could loose this requirement and let the developer to decide if he is going to imlement the `Display` for the _kind_ or for the whole _error_ only. Only if he implements the `Display` for the kind, the error would be displayed as well _out of the box_, otherwise it should be implemented for the specific error to satisfy its `std::error::Error` trait.

Ah, yes, you're right! The context is available, though printed as
```
failed publishing object chunks:: failed chunk publish: something
```

so all `:` at the end at the errors should go away. But rather in in-code impls, not in generic error impl.

Let's get it merged and discuss potential further improvements outside of this PR.

nepalez commented:
Yeah, my bad. Fixed (removed excessive `:`)

Jarema reviewed (approved):
LGTM!
